unknown_type/[M]378.cpp

```diff
@@ -0,0 +1,20 @@
+// Problem: 378. Kth Smallest Element in a Sorted Matrix - Medium
+// Link: https://leetcode.com/problems/kth-smallest-element-in-a-sorted-matrix/
+
+/* Nhận xét:
+- Bài này anh chịu rồi :))
+*/
+
+class Solution {
+public:
+    int kthSmallest(vector<vector<int>>& matrix, int k) {
```

Comment:
Uh code O(M*N * Log(M*N)) thế này cũng được. Có cách code độ phức tạp là O(M*N* LOG(K)), chú có muốn thử ko? Hint: Dùng heap.

Reply:
Anh cũng có một chút mường tượng: kiểu như loang từ phần tử nhỏ nhất, add tất cả các đỉnh kề vào heap. xong lấy lần lượt các phần tử đầu ra khỏi heap. Lấy đủ K phần tử thì thôi. Vì số lượng các phần tử nằm trong heap luôn khá nhỏ độ phức tạp sẽ không tới log(m*n).
Nhưng căn bản là cứ nghĩ tới phương pháp hash_table hoặc binary search nên cảm giác bế tắc thực sự :(

Reply:
Ý tưởng đúng rồi đấy. Code thử xem. Cho quen trước với heap cũng được :D

Reply:
- Bài này sau khi sửa theo C2 thì thời gian chạy thậm chỉ còn tăng lên, có vẻ như overhead là quá lớn. (C1: 32ms vs C2: 52ms). Chú nghĩ sao?

Reply:
Chú thử code theo cách: BƯớc đầu thả N thằng đầu của các hàng vào xem, tại mỗi bước khi lấy ra khỏi heap, thả tiếp vào heap thằng ngay bên cạnh nó. Code sẽ đẹp hơn.

Reply:
Thanks for your hint :) Anh update code theo hướng dẫn của chú rồi, code đẹp hơn nhiều, chú kiểm tra lại nhé.
Cơ mà khi submit thì thời gian chạy gần như không thay đổi nhiều, vẫn chậm hơn cách đầu tiên khá nhiều (48ms vs 32ms)

Reply:
@quynhtruong anh vừa update, chú kiểm tra lại nhé.
Sau khi sửa code, thì thời gian có vẻ đã giảm đi, dao động từ 36ms->44ms
Tuy nhiên so với cách đơn giản anh làm, tức là sort xong lấy giá trị, thì phương pháp đó chỉ có thời gian chạy ổn định ở mức: 28ms -> 32ms.
Tức là anh thấy mình đang code phức tạp hơn chút mà thời gian thực thi có vẻ lại tăng lên.
Chú tính sao?